### DISCUSSION TITLE
@t.mongoId - what should it do?

### DISCUSSION THREAD

**Discussion by smolinari** [marcj/deepkit#143] [Q&A] [answered]
Hi,

What is the purpose of `@t.mongoId`? I was expecting it to return a string when used with `classToPlain`, but it is returning `new objectId('*a mongoId here!*")`.

Scott

**Comment by marcj** [comment, marked as answer]
It will be a string. It can't return an ObjectId because we don't have such a class in our codebase. If you use mongo's official client, then this won't work, as "t.mongoId" is for deepkit/orm, not mongo's javascript driver. If you pass in an object from mongo's driver, it probably won't do anything with its type, which is what you see. What does your code look like?

**Reply by smolinari** [reply]
We're working with Mongoose. I've created a gist for the tests we created. I sort of tricked myself with the first four tests by assigning the (incorrect) id to the comparison object (`accountClassObject`) and the maintainer of Typegoose caught the type problem, but a bit by accident, as he was trying to make sure the same issue doesn't happen that has happened with `class-transformer` (and why we started this endeavor with `@deepkit/type`). The `class-transformer` issue is a different one though.  

https://gist.github.com/smolinari/17b769b81bb0f7964764ea93abcd69b8

The 5th test is from the Typegoose author, where he found the problem. The last one, the 6th, is from me again. 

If you want to look at the whole repo, you can see it here (minus my 6th test, as I've not committed and pushed it yet): https://github.com/smolinari/typegoose

- Switch to the beta branch.
- Run `yarn`
- Run `yarn run test:watch deepkitType`

Notice the last line and comment of the gist. I appreciate your insights and help. Thanks!

By the way, not sure if you want more publicity for your framework/ library, but because it seems `class-transformer` is abandoned, more people will be looking for a de-/serialization library like yours and Mongoose has a huge usership. This could be just the start of a lot of attention, despite your framework not using Mongoose. 😉 

Scott

**Reply by marcj** [reply]
I don't understand what you are saying. deepkit/type's mongoId does not work out of the box with all kind of libraries like Mongoose. It does not even work with Mongo's node-mongo-driver. That being said, you can certainly (and deepkit/type is designed for that) to customize the serializer to do that for you. See https://deepkit.io/documentation/type/serialization-target

```typescript
    import {jsonSerializer} from '@deepkit/type';
    import {ObjectId} from 'bson';

    const mySerializer = jsonSerializer.fork('mySerializer');

    //We overwrite mongoId and correctly convert from Mongo ObjectID to string when deserializing
    mySerializer.toClass.register('objectId', (property, state) => {
        state.setContext({ObjectId: ObjectId});
        state.addSetter(`${state.accessor} instanceof ObjectId ? ${state.accessor}.toHexString() : ${state.accessor}`);
    });

    //We overwrite mongoId and correctly convert string to Mongo ObjectID when serializing
    mySerializer.fromClass.register('objectId', (property, state) => {
        state.setContext({ObjectId: ObjectId});
        state.addSetter(`${state.accessor} instanceof ObjectId ? ${state.accessor} : ObjectId.createFromHexString(${state.accessor})`);
    });
```

and then use this mySerializer instead of classToPlain/plainToClass (which use jsonSerializer per default):

```typescript
class MyModel {
     @t.mongoId id?: string;
}

const m = new MyModel;
m.id = '507f191e810c19729de860ea';

mySerializer.for(MyModel).serialize(m); // {id: ObjectId}
mySerializer.for(MyModel).deserialize({id: new ObjectId('507f191e810c19729de860ea')); //{id: '507f191e810c19729de860ea'}
```

Note: that code is only from the top of my head, so no guarantee that it works or even compiles. But I hope you get the idea.

**Reply by smolinari** [reply]
Wow, that was a fast reply. I was still editing my post, and you had already replied. Thanks! 

Yeah. I understand now the `@t.mongoId` is specialized. I guess I was a bit too optimistic. :)

I'll take you up on the override and see how it goes. Appreciate the help!

Scott

**Reply by smolinari** [reply]
@marcj - Ok. I tried your suggestion. But, I ran into type errors with `jsonSerializer.fork`. 

![image](https://user-images.githubusercontent.com/3314957/132158175-739b6c47-0e17-4b26-a7a4-8a04a09dafd8.png)

I also tried just creating a new Serializer. I don't have the type issues, but then I get the same results as if nothing was overridden. If I create the new Serializer and then fork it, I also get the type errors. 

Scott

**Reply by marcj** [reply]
See https://github.com/deepkit/deepkit-framework/blob/master/packages/type/tests/serializer-api.spec.ts#L91 for examples.

As I said, I've posted previously examples that are not necessarily correct. But in the tests you find working examples. 

It should be

```typescript
    const mySerializer = new class extends jsonSerializer.fork('mySerializer') {
    };

//or

    const mySerializer = new jsonSerializer.fork('mySerializer');
```

**Reply by smolinari** [reply]
Ok. That helped. Thanks so much!

I also found the logic was backwards. `toClass` (deserialization) should be creating the ObjectId. `fromClass` (serialization) the string. 😁 

But, awesome stuff and all is working now as expected!

Thanks again. 👍 

Scott
